PWGJE/Core/JetFinder.cxx

```diff
@@ -14,82 +14,49 @@
-  jetDefBkg = fastjet::JetDefinition(algorithmBkg, jetBkgR, recombSchemeBkg, strategyBkg);
-  areaDefBkg = fastjet::AreaDefinition(areaTypeBkg, ghostAreaSpec);
-  selRho = fastjet::SelectorRapRange(bkgEtaMin, bkgEtaMax) && fastjet::SelectorPhiRange(bkgPhiMin, bkgPhiMax); //&& !fastjet::SelectorNHardest(2)    //here we have to put rap range, to be checked!
-}
 
-/// Sets the background subtraction estimater pointer
-void JetFinder::setBkgE()
-{
-  if (bkgSubMode == BkgSubMode::rhoAreaSub || bkgSubMode == BkgSubMode::constSub) {
-    bkgE = decltype(bkgE)(new fastjet::JetMedianBackgroundEstimator(selRho, jetDefBkg, areaDefBkg));
-  } else {
-    if (bkgSubMode != BkgSubMode::none) {
-      LOGF(error, "requested subtraction mode not implemented!");
-    }
-  }
-}
-
-/// Sets the background subtraction pointer
-void JetFinder::setSub()
-{
-  //if rho < 1e-6 it is set to 1e-6 in AliPhysics
-  if (bkgSubMode == BkgSubMode::rhoAreaSub) {
-    sub = decltype(sub){new fastjet::Subtractor{bkgE.get()}};
-  } else if (bkgSubMode == BkgSubMode::constSub) { //event or jetwise
-    constituentSub = decltype(constituentSub){new fastjet::contrib::ConstituentSubtractor{bkgE.get()}};
-    constituentSub->set_distance_type(fastjet::contrib::ConstituentSubtractor::deltaR);
-    constituentSub->set_max_distance(constSubRMax);
-    constituentSub->set_alpha(constSubAlpha);
-    constituentSub->set_ghost_area(ghostArea);
-    constituentSub->set_max_eta(bkgEtaMax);
-    constituentSub->set_background_estimator(bkgE.get()); //what about rho_m
-  } else {
-    if (bkgSubMode != BkgSubMode::none) {
-      LOGF(error, "requested subtraction mode not implemented!");
-    }
-  }
+  mSubUtils = std::make_unique<JetBkgSubUtils>(jetR, bkgPhiMin, bkgPhiMax, bkgEtaMin, bkgEtaMax, constSubAlpha, constSubRMax, ghostAreaSpec);
+  mSubUtils->setJetAlgorithmAndScheme(algorithmBkg, recombSchemeBkg);
 }
 
 /// Performs jet finding
 /// \note the input particle and jet lists are passed by reference
 /// \param inputParticles vector of input particles/tracks
 /// \param jets veector of jets to be filled
 /// \return ClusterSequenceArea object needed to access constituents
-fastjet::ClusterSequenceArea JetFinder::findJets(std::vector<fastjet::PseudoJet>& inputParticles, std::vector<fastjet::PseudoJet>& jets) //ideally find a way of passing the cluster sequence as a reeference
+fastjet::ClusterSequenceArea JetFinder::findJets(std::vector<fastjet::PseudoJet>& inputParticles, std::vector<fastjet::PseudoJet>& jets) // ideally find a way of passing the cluster sequence as a reeference
 {
   setParams();
-  setBkgE();
   jets.clear();
 
-  if (bkgE) {
-    bkgE->set_particles(inputParticles);
-    setSub();
-  }
-  if (constituentSub) {
-    inputParticles = constituentSub->subtract_event(inputParticles);
-  }
+  sub = mSubUtils->setSub(inputParticles, mRho, bkgSubMode);
```

Comment:
you can put a comment here saying in case of constituentsub the subtraction is already done here in this line. We can also add an extra constituent sub option (separate to this one) like there used to be where the constituent subtraction is only applied to the clustered jet constituents and not to the whole event. This could be implemented like the perpendicular cone method which takes in the jets

Reply:
I see, yes I will do it ASAP.